### PR TITLE
PoA Register Authority Command

### LINKED ISSUE #8413: Implement events for Proof of Authority module
### Description

Add events definitions as defined in the [LIP](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0047.md).

### Acceptance Criteria

- All events described in the LIP must be implemented

### LINKED ISSUE #8414: Implement Register Authority command for PoA module
### Description

Implement register authority command as described in the [LIP](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0047.md#register-authority-command)

Add dependencies used in command as necessary.

### Acceptance Criteria

- Add unit tests covering all the path.

### PR DESCRIPTION
**Pull request by Phanco**
### What was the problem?

This PR resolves #8414

### How was it solved?
Implemented RegisterAuthorityCommand

### How was it tested?
Unit Test added

### CODE REVIEW

File: framework/src/modules/poa/commands/register_authority.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		try {
+			validator.validate(registerAuthorityParamsSchema, context.params);
+		} catch (err) {
+			return {
+				status: VerifyStatus.FAIL,
+				error: err as Error,
+			};
+		}
+
+		if (!/^[a-z0-9!@$&_.]+$/g.test(name)) {
```
[the remark this answers is not in the document]
Comment (Phanco) [reply framework/src/modules/poa/commands/register_authority.ts:56]: empty names will be blocked by `validator.validate`

**Review by ishantiw — CHANGES_REQUESTED** [review]

File: framework/src/modules/poa/commands/register_authority.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+	public async execute(context: CommandExecuteContext<RegisterAuthorityParams>): Promise<void> {
+		const { params } = context;
+
+		const senderAddress = getSenderAddress(context.transaction.senderPublicKey);
+		this._feeMethod.payFee(context, REGISTRATION_FEE);
+
+		await this.stores.get(ValidatorStore).set(context, senderAddress, {
+			name: params.name,
+		});
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/poa/commands/register_authority.ts:84]: encoding is already done is `BaseStore` class `setWithSchema`

File: framework/src/modules/poa/commands/register_authority.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+		await this.stores.get(NameStore).set(context, Buffer.from(params.name), {
+			address: senderAddress,
+		});
```
[the remark this answers is not in the document]
Comment (mitsuaki-u) [reply framework/src/modules/poa/commands/register_authority.ts:88]: same as above

File: framework/test/unit/modules/poa/commands/register_authority.spec.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	const transactionParams = {
+		name: 'max',
+		blsKey: utils.getRandomBytes(48),
+		proofOfPossession: utils.getRandomBytes(96),
+		generatorKey: utils.getRandomBytes(32),
+	};
+
+	const publicKey = utils.getRandomBytes(32);
+	const chainID = Buffer.from(
+		'e48feb88db5b5cf5ad71d93cdcd1d879b6d5ed187a36b0002cc34e0ef9883255',
```
[the remark this answers is not in the document]
Comment (Phanco) [reply framework/test/unit/modules/poa/commands/register_authority.spec.ts:63]: It's just random string, changed to use `utils.getRandomBytes(32)`

File: framework/src/modules/poa/commands/register_authority.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Phanco) [reply framework/src/modules/poa/commands/register_authority.ts:84]: The encoding is performed inside `ValidatorStore`

File: framework/src/modules/poa/types.ts [hunk 5]
```diff
@@ -11,3 +11,52 @@
...
+	setValidatorGeneratorKey(
+		methodContext: MethodContext,
+		validatorAddress: Buffer,
+		generatorKey: Buffer,
+	): Promise<boolean>;
+	// https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#registervalidatorkeys
+	registerValidatorKeys(
+		methodContext: MethodContext,
+		validatorAddress: Buffer,
+		blsKey: Buffer,
```
Comment (sergeyshemyakov) [inline framework/src/modules/poa/types.ts:37]: Note that [the LIP](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#registervalidatorkeys) gives the following order of parameters: `validatorAddress, proofOfPossession, generatorKey, blsKey` which is different from the order that you mention in the interface here. Please make sure that your code is consistent with the SDK (no need to be consistent with the LIP IMO).

File: framework/test/unit/modules/poa/commands/register_authority.spec.ts [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+	const registerAuthorityTransactionParams = {
+		name: 'max',
+		blsKey: utils.getRandomBytes(LENGTH_BLS_KEY),
+		proofOfPossession: utils.getRandomBytes(LENGTH_PROOF_OF_POSSESSION),
+		generatorKey: utils.getRandomBytes(LENGTH_GENERATOR_KEY),
+	};
+
+	const publicKey = utils.getRandomBytes(32);
+	const chainID = utils.getRandomBytes(32);
```
Comment (sergeyshemyakov) [inline framework/test/unit/modules/poa/commands/register_authority.spec.ts:62]: According to the protocol, chain ID must have length 4 bytes.

**Review by sergeyshemyakov — COMMENTED** [review]
I have just two small comments, generally the PR is really good!

File: framework/src/modules/poa/types.ts (hunk 5 above)
Comment (Phanco) [reply framework/src/modules/poa/types.ts:37]: You are right. However on the validators module it is arranged that way. Maybe we should open a new issue to re-align this function to the LIP

Comment (sergeyshemyakov) [reply framework/src/modules/poa/types.ts:37]: I see no problem if the implementation differs from the LIP in this small way, the important bit is that the protocol is exactly the same.

**Review by sergeyshemyakov — APPROVED** [review]
Looks good to me!

**Review by mitsuaki-u — APPROVED** [review]

**Review by ishantiw — APPROVED** [review]

**Pull request merged**